File: app/src/main/java/com/ilyap/yuta/ui/adapters/BaseAdapter.java
package com.ilyap.yuta.ui.adapters;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import lombok.AllArgsConstructor;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;

import java.util.List;

@Getter
@AllArgsConstructor
public abstract class BaseAdapter<T, VH extends BaseAdapter.ViewHolder<T>> extends RecyclerView.Adapter<VH> implements ListAdapterUpdater<T> {
    private final Context context;
    private final List<T> items;

    @Override
    public void onBindViewHolder(@NotNull VH holder, int position) {
        holder.bind(items.get(position));
    }

    @SuppressLint("NotifyDataSetChanged")
    @Override
    public void updateList(List<T> items) {
        this.items.clear();
        this.items.addAll(items);
        this.notifyDataSetChanged();
    }

    public void removeItem(T item) {
        int position = this.items.indexOf(item);
        this.notifyItemRemoved(position);
        this.items.remove(position);
        this.notifyItemRangeChanged(position, getItemCount());
    }

    public void insertItem(T item) {
        this.items.add(item);
        int position = this.items.indexOf(item);
        this.notifyItemInserted(position);
        this.notifyItemRangeChanged(position, getItemCount());
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public abstract static class ViewHolder<T> extends RecyclerView.ViewHolder {
        public ViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        public abstract void bind(T item);
    }
}

File: app/src/main/java/com/yuta/common/ui/BaseAdapter.kt
package com.yuta.common.ui

import android.view.View
import androidx.recyclerview.widget.RecyclerView

abstract class BaseAdapter<T, VH : BaseAdapter.ViewHolder<T>>(
    val items: MutableList<T>
) : RecyclerView.Adapter<VH>() {

    override fun onBindViewHolder(holder: VH, position: Int) {
        holder.bind(items[position])
    }

    @Suppress("NotifyDataSetChanged")
    fun refillList(newItems: List<T>) {
        items.clear()
        items.addAll(newItems)
        notifyDataSetChanged()
    }

    fun removeItem(item: T) {
        val position = items.indexOf(item)
        if (position != -1) {
            items.removeAt(position)
            notifyItemRemoved(position)
            notifyItemRangeChanged(position, itemCount)
        }
    }

    fun insertItem(item: T) {
        items.add(item)
        val position = items.indexOf(item)
        notifyItemInserted(position)
        notifyItemRangeChanged(position, itemCount)
    }

    override fun getItemCount(): Int = items.size

    abstract class ViewHolder<T>(itemView: View) : RecyclerView.ViewHolder(itemView) {
        abstract fun bind(item: T)
    }
}
